lib/ecto/type.ex

```diff
@@ -1387,6 +1387,13 @@ defmodule Ecto.Type do
     {:ok, Enum.reverse(acc)}
   end
 
+  defp array_with_index(%_{} = struct, fun, skip_nil?, index, acc) do
+    case Enumerable.impl_for(struct) do
+      nil -> :error
+      _ -> struct |> Enum.to_list() |> array_with_index(fun, skip_nil?, index, acc)
+    end
+  end
+
   defp array_with_index(_, _, _, _, _) do
     :error
   end
```

Comment:
```suggestion
  defp array_with_index(other, fun, skip_nil?, index, acc) do
    case Enumerable.impl_for(other) do
      nil -> :error
      _ -> other |> Enum.to_list() |> array_with_index(fun, skip_nil?, index, acc)
    end
  end
```
@josevalim is there any reason to limit it to structs or should we do something like this?

Reply:
The only other built in is maps and I don’t think it would work here?

Reply:
ah because it turns into a keyword list right? yeah makes sense!

Reply:
I updated the documentation specifically calling out structs, as I'm not sure that comparing the key/value pairs may work intuitively?

Reply:
We could pull out the values of a keyword/map and call it out in the documentation, as well.

Reply:
imo it would be a bit confusing to do something like this `from t in Table, where: t.int_row in %{a: 1, b: 2}`. to me the current change looks good. will approve and if jose agrees to leave out maps/kw lists then we are good to go